### PR TITLE
Adding streaming support to the benchmarking script

### PR DESCRIPTION
**Pull request by Edwinhr716**
Added new variable `stream_request` to determine whether or not to stream the request. Default value is false. Only vLLM is supported.

Also added TTFT metric.

### CODE REVIEW

File: benchmarks/benchmark/tools/profile-generator/container/benchmark_serving.py [hunk 1]
```diff
@@ -572,6 +676,17 @@ def print_and_save_result(args: argparse.Namespace, benchmark_duration, total_re
...
+  if args.stream_request:
+    mean_ttft=np.mean(ttfts)
+    median_ttft=np.median(ttfts)
+    p99_ttft=np.percentile(ttfts, 99)
+    print(f"Mean TTFT (s): {mean_ttft:.2f}")
+    print(f"Median TTFT (s): {median_ttft:.2f}")
+    print(f"P99 TTFT (s): {p99_ttft:.2f}")
+    benchmark_result['mean_ttft'] = mean_ttft
+    benchmark_result['median_ttft'] = median_ttft
+    benchmark_result['p99_ttft'] = p99_ttft
```
Comment (Bslabe123) [inline benchmarks/benchmark/tools/profile-generator/container/benchmark_serving.py:688]: nit, why not use `get_stats_for_set()`

File: benchmarks/benchmark/tools/profile-generator/container/benchmark_serving.py [hunk 2]
```diff
@@ -321,21 +406,40 @@ async def benchmark(
...
+          prompt_len,
+          output_len,
+          args.best_of,
+          args.use_beam_search,
+          args.top_k,
+          tokenizer,
+          args.sax_model,
+          model,
+        )
+      )
```
Comment (Bslabe123) [inline benchmarks/benchmark/tools/profile-generator/container/benchmark_serving.py:425]: Something like functools.partial may make this cleaner, but not blocking

File: benchmarks/benchmark/tools/profile-generator/container/benchmark_serving.py [hunk 3]
```diff
@@ -134,6 +134,90 @@ def init_errors_map() -> Dict[str, int]:
...
+        "best_of": best_of,
+        "use_beam_search": use_beam_search,
+        "temperature": 0.0 if use_beam_search else 1.0,
+        "top_p": 1.0,
+        "max_tokens": output_len,
+        "ignore_eos": True,
+        "stream": True,
+    }
+  else: 
+    raise ValueError(f"Unknown backend: {backend}")
```
Comment (Bslabe123) [inline benchmarks/benchmark/tools/profile-generator/container/benchmark_serving.py:170]: How certain are we that streaming isn't supported in jetstream?

Comment (achandrasekar) [reply benchmarks/benchmark/tools/profile-generator/container/benchmark_serving.py:170]: Streaming with gRPC is supported. I believe HTTP is not yet and @vivianrwu is working on it.

File: benchmarks/benchmark/tools/profile-generator/container/benchmark_serving.py [hunk 4]
```diff
@@ -134,6 +134,90 @@ def init_errors_map() -> Dict[str, int]:
...
+    tokenizer: PreTrainedTokenizerBase,
+    sax_model: str,
+    model: str,
+) -> Tuple[Tuple[int, int, float], float, Dict[str, int]]:
+  """Sends stream request to server"""
+  request_start_time = time.time()
+  errors = init_errors_map()
+
+  headers = {"User-Agent": "Benchmark Client"}
+  # Doesn't work with if model weights are stored in PD on the model server side
```
Comment (achandrasekar) [inline benchmarks/benchmark/tools/profile-generator/container/benchmark_serving.py:155]: Why is this the case?

**Review by achandrasekar — COMMENTED** [review]
Would be good to see an example of the output json as well with streaming to see how it looks.

File: benchmarks/benchmark/tools/profile-generator/container/benchmark_serving.py (hunk 2 above)
Comment (Edwinhr716) [reply benchmarks/benchmark/tools/profile-generator/container/benchmark_serving.py:424]: Mmm functools.partial would require only using one function for both streaming and non-streaming requests, and hardcoding one of the values with partial. I think having send_request and send_request_streaming be the same function with different implementations allows for easier refactoring in the future. What do you think?

File: benchmarks/benchmark/tools/profile-generator/container/benchmark_serving.py (hunk 4 above)
Comment (Edwinhr716) [reply benchmarks/benchmark/tools/profile-generator/container/benchmark_serving.py]: Oops forgot to remove this comment. This was before https://github.com/GoogleCloudPlatform/ai-on-gke/commit/a3401f29d0110a8d62dccd6ff1f5217f7766b1a3 was merged. Before, tokenizer was used for both the tokenizer and the model. So if the model was downloaded in a PD and stored in a directory not called for instance `meta-llama/Llama-3.1-405B-Instruct `,  the model server would not find the model.

File: benchmarks/benchmark/tools/profile-generator/container/benchmark_serving.py (hunk 1 above)
Comment (Edwinhr716) [reply benchmarks/benchmark/tools/profile-generator/container/benchmark_serving.py]: No particular reason. Changed it to use `get_stats_for_set()` instead.

**Comment by Edwinhr716** [comment]
This is what the output of a normal run looks like 
```
Total time: 95.20 s
Successful/total requests: 60/60
Requests/min: 37.81
Output_tokens/min: 4002.02
Input_tokens/min: 2626.21
Tokens/min: 6628.23
Average Time to First Token (s): 0.44
Average seconds/token (includes waiting time on server): 0.04
Average milliseconds/request (includes waiting time on server): 6284.16
Average milliseconds/output_token (includes waiting time on server): 65.38
Average input length: 69.45
Average output length: 105.83
```

And the output json 
```
{
  "metrics": {
    "num_prompts_attempted": 60,
    "num_prompts_succeeded": 60,
    "request_rate": 1,
    "server_metrics": {},
    "benchmark_time": 95.20189309120178,
    "throughput_rps": 0.6302395682669989,
    "throughput": 66.70035430825739,
    "total_output_token": 6350,
    "output_tokens_per_min": 4002.0212584954434,
    "total_input_tokens": 4167,
    "input_tokens_per_min": 2626.2082809685844,
    "total_tokens": 10517,
    "tokens_per_min": 6628.229539464028,
    "avg_per_token_latency": 0.03693258923504509,
    "median_per_token_latency": 0.04176087070985145,
    "sd_per_token_latency": 0.01983002434856721,
    "min_per_token_latency": 0.003826681871591888,
    "max_per_token_latency": 0.059450237118467995,
    "p90_per_token_latency": 0.0565482251304172,
    "p99_per_token_latency": 0.05912417530913509,
    "avg_TTFT": 0.43885604144694906,
    "median_TTFT": 0.43761963001452386,
    "sd_TTFT": 0.027700201056499105,
    "min_TTFT": 0.38527003629133105,
    "max_TTFT": 0.5583119788207114,
    "p90_TTFT": 0.4659155633300543,
    "p99_TTFT": 0.5339873587712645,
    "avg_latency": 6284.161829948425,
    "median_latency": 5671.264052391052,
    "sd_latency": 4454.538686914545,
    "min_latency": 616.095781326294,
    "max_latency": 15108.333587646484,
    "p90_latency": 12943.064212799072,
    "p99_latency": 14828.856890201567,
    "avg_per_output_token_latency": 65.3789068777566,
    "median_per_output_token_latency": 60.812312981178025,
    "sd_per_output_token_latency": 12.65093424415054,
    "min_per_output_token_latency": 53.26013545277702,
    "max_per_output_token_latency": 123.21915626525879,
    "p90_per_output_token_latency": 75.11993244890766,
    "p99_per_output_token_latency": 110.71500478472021,
    "avg_input_len": 69.45,
    "median_input_len": 26,
    "sd_input_len": 75.60697168030649,
    "min_input_len": 5,
    "max_input_len": 255,
    "p90_input_len": 199.70000000000002,
    "p99_input_len": 254.41,
    "avg_output_len": 105.83333333333333,
    "median_output_len": 92.5,
    "sd_output_len": 79.34233142248566,
    "min_output_len": 5,
    "max_output_len": 257,
    "p90_output_len": 229.3,
    "p99_output_len": 253.45999999999998,
    "ClientConnectorError": 0,
    "TimeoutError": 0,
    "ContentTypeError": 0,
    "ClientOSError": 0,
    "ServerDisconnectedError": 0,
    "unknown_error": 0
  },
  "dimensions": {
    "date": "20241101-214038",
    "backend": "vllm",
    "model_id": "meta-llama/Meta-Llama-3.1-405B-Instruct",
    "tokenizer_id": "meta-llama/Meta-Llama-3.1-405B-Instruct"
  },
  "config": {
    "model": "meta-llama/Meta-Llama-3.1-405B-Instruct",
    "num_models": 1,
    "model_server": "vllm",
    "start_time": {
      "seconds": 1730497238,
      "nanos": 811286000
    }
  },
  "summary_stats": {
    "stats": [
      {
        "request_rate": 1,
        "request_latency": {
          "mean": 6284.161829948425,
          "median": 5671.264052391052,
          "sd": 4454.538686914545,
          "min": 616.095781326294,
          "max": 15108.333587646484,
          "p90": 12943.064212799072,
          "p99": 14828.856890201567
        },
        "throughput": {
          "mean": 66.70035430825739
        },
        "input_length": {
          "mean": 69.45,
          "median": 26,
          "sd": 75.60697168030649,
          "min": 5,
          "max": 255,
          "p90": 199.70000000000002,
          "p99": 254.41
        },
        "output_length": {
          "mean": 105.83333333333333,
          "median": 92.5,
          "sd": 79.34233142248566,
          "min": 5,
          "max": 257,
          "p90": 229.3,
          "p99": 253.45999999999998
        },
        "tpot": {
          "mean": 65.3789068777566,
          "median": 60.812312981178025,
          "sd": 12.65093424415054,
          "min": 53.26013545277702,
          "max": 123.21915626525879,
          "p90": 75.11993244890766,
          "p99": 110.71500478472021
        },
        "model_server_metrics": []
      }
    ]
  }
}

```

**Review by Bslabe123 — APPROVED** [review]

**Review by achandrasekar — APPROVED** [review]

**Pull request merged**